edgelet/iotedge/src/support_bundle.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::fs::File;
+use std::path::Path;
+
+use chrono::{DateTime, Local, NaiveDateTime, Utc};
+use failure::Fail;
+use futures::{Future, Stream};
+use tokio::prelude::*;
+use zip;
+
+use crate::error::{Error, ErrorKind};
+use edgelet_core::{LogOptions, LogTail, Module, ModuleRuntime};
+
+use crate::logs::pull_logs;
+use crate::Command;
+
+pub struct SupportBundle<M> {
+    runtime: M,
+    log_options: LogOptions,
+    location: String,
+}
+
+struct BundleState<M> {
+    runtime: M,
+    log_options: LogOptions,
+    location: String,
+    file_options: zip::write::FileOptions,
+    zip_writer: zip::ZipWriter<File>,
+}
+
+impl<M> Command for SupportBundle<M>
+where
+    M: 'static + ModuleRuntime + Clone + Send + Sync,
+{
+    type Future = Box<dyn Future<Item = (), Error = Error> + Send>;
+
+    fn execute(self) -> Self::Future {
+        let result = future::result(self.make_state())
+            .and_then(SupportBundle::write_all_logs)
+            .map(drop)
+            .map(|_| println!("Wrote all logs to file"));
+
+        Box::new(result)
+    }
+}
+
+impl<M> SupportBundle<M>
+where
+    M: 'static + ModuleRuntime + Clone + Send + Sync,
+{
+    pub fn new(log_options: LogOptions, location: &str, runtime: M) -> Self {
+        SupportBundle {
+            runtime,
+            log_options,
+            location: location.to_owned(),
+        }
+    }
+
+    fn make_state(self) -> Result<BundleState<M>, Error> {
+        /* Print status */
+        let since_time: DateTime<Utc> = DateTime::from_utc(
+            NaiveDateTime::from_timestamp(self.log_options.since().into(), 0),
+            Utc,
+        );
+        let since_local: DateTime<Local> = DateTime::from(since_time);
+        let max_lines = if let LogTail::Num(tail) = self.log_options.tail() {
+            format!("(maximum {} lines) ", tail)
+        } else {
+            "".to_owned()
+        };
+        println!(
+            "Writing all logs {}since {} (local time {}) to {}",
+            max_lines, since_time, since_local, self.location
+        );
+
+        /* Make state */
+        let file_options =
+            zip::write::FileOptions::default().compression_method(zip::CompressionMethod::Deflated);
+
+        let zip_writer = zip::ZipWriter::new(
+            File::create(format!("{}/css_bundle.zip", self.location.to_owned()))
+                .map_err(|err| Error::from(err.context(ErrorKind::WriteToStdout)))?,
```

Comment:
We should probably add a different `ErrorKind` than `WriteToStdout`. This is going to possibly print an error message that doesn't make sense because this is not writing to stdout.